File: src/main/java/dev/wbell/terrariateleporter/waystonePosition.java
package dev.wbell.terrariateleporter;

import com.google.gson.Gson;
import org.bukkit.plugin.java.JavaPlugin;

import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;


public class waystonePosition {

    private static final Gson gson = new Gson();
    private static File dataFile = new File("");
    public static final List<PositionData> positions = new ArrayList<PositionData>();
    public static PositionData waystoneNear(PositionData pos) {
        for (int height = -1; height <= 2; height++) {
            for (int width = -1; width <= 1; width++) {
                for (int length = -1; length <= 1; length++) {
                    if (waystoneExists(new PositionData(pos.getX() + width, pos.getY() + height, pos.getZ() + length))) {
                        return new PositionData(pos.getX() + width, pos.getY() + height, pos.getZ() + length);
                    }
                }
            }
        }
        for (PositionData position : positions) {
            if (position.getX() == pos.getX() && position.getY() == pos.getY() && position.getZ() == pos.getZ()) {
                return position;
            }
        }
        return null;
    }
    public static boolean waystoneExists(PositionData pos) {
        for (PositionData position : positions) {
            if (position.getX() == pos.getX() && position.getY() == pos.getY() && position.getZ() == pos.getZ()) {
                return true;
            }
        }
        return false;
    }

    public static void addWaystone(PositionData position) {
        positions.add(position);
        savePositions();
    }

    public static void removeWaystone(PositionData position) {
        for (int i = 0; i < positions.size(); i++) {
            PositionData pos = positions.get(i);
            if (position.getX() == pos.getX() && position.getY() == pos.getY() && position.getZ() == pos.getZ()) {
                positions.remove(i);
                return;
            }
        }
        savePositions();
    }

    public static List<PositionData> getAllPositionNotIncluding(PositionData position) {
        List<PositionData> positionsNotIncluding = new ArrayList<PositionData>();
        for (PositionData pos : positions) {
            if (position.getX() == pos.getX() && position.getY() == pos.getY() && position.getZ() == pos.getZ())
                continue;
            positionsNotIncluding.add(pos);
        }
        return positionsNotIncluding;
    }

    public List<PositionData> getPositions() {
        return positions;
    }

    public static void loadPositions(File file) {
        dataFile = file;
        if (!file.exists()) {
            savePositions();
            return;
        }

        try (Reader reader = new FileReader(file)) {
            PositionData[] loadedPositions = gson.fromJson(reader, PositionData[].class);
            positions.clear();
            positions.addAll(Arrays.asList(loadedPositions));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void savePositions() {
        try (Writer writer = new FileWriter(dataFile)) {
            gson.toJson(positions.toArray(new PositionData[0]), writer);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/java/dev/wbell/terrariateleporter/waystonePosition.kt
package dev.wbell.terrariateleporter

import com.google.gson.Gson
import java.io.File
import java.io.FileReader
import java.io.FileWriter
import java.io.IOException
import java.util.*

class waystonePosition {
    val positions: List<PositionData>
        get() = Companion.positions

    companion object {
        private val gson = Gson()
        private var dataFile: File? = null
        val positions: MutableList<PositionData> = ArrayList()
        fun waystoneNear(pos: PositionData): PositionData? {
            for (height in -1..2) {
                for (width in -1..1) {
                    for (length in -1..1) {
                        if (waystoneExists(PositionData(pos.x + width, pos.y + height, pos.z + length))) {
                            return PositionData(pos.x + width, pos.y + height, pos.z + length)
                        }
                    }
                }
            }
            for (position in positions) {
                if (position.x == pos.x && position.y == pos.y && position.z == pos.z) {
                    return position
                }
            }
            return null
        }

        fun waystoneExists(pos: PositionData): Boolean {
            for (position in positions) {
                if (position.x == pos.x && position.y == pos.y && position.z == pos.z) {
                    return true
                }
            }
            return false
        }

        fun addWaystone(position: PositionData) {
            positions.add(position)
            savePositions()
        }

        fun removeWaystone(position: PositionData) {
            for (i in positions.indices) {
                val pos = positions[i]
                if (position.x == pos.x && position.y == pos.y && position.z == pos.z) {
                    positions.removeAt(i)
                    return
                }
            }
            savePositions()
        }

        fun getAllPositionNotIncluding(position: PositionData): List<PositionData> {
            val positionsNotIncluding: MutableList<PositionData> = ArrayList()
            for (pos in positions) {
                if (position.x == pos.x && position.y == pos.y && position.z == pos.z) continue
                positionsNotIncluding.add(pos)
            }
            return positionsNotIncluding
        }

        fun loadPositions(file: File) {
            dataFile = file
            if (!file.exists()) {
                savePositions()
                return
            }
            try {
                FileReader(file).use { reader ->
                    val loadedPositions = gson.fromJson(reader, Array<PositionData>::class.java)
                    positions.clear()
                    positions.addAll(listOf(*loadedPositions))
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }

        fun savePositions() {
            try {
                dataFile?.let { FileWriter(it).use { writer -> gson.toJson(positions.toTypedArray<PositionData>(), writer) } }
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }
}